homeassistant/components/zeroconf/__init__.py

```diff
@@ -38,19 +39,26 @@ def setup(hass, config):
 
     try:
         host_ip_pton = socket.inet_pton(socket.AF_INET, host_ip)
+        info = ServiceInfo(ZEROCONF_TYPE, zeroconf_name, address=host_ip_pton,
+                           port=hass.http.server_port, weight=0, priority=0,
+                           properties=params)
     except socket.error:
         host_ip_pton = socket.inet_pton(socket.AF_INET6, host_ip)
+        info = ServiceInfo(ZEROCONF_TYPE, zeroconf_name, address6=host_ip_pton,
+                           port=hass.http.server_port, weight=0, priority=0,
+                           properties=params)
 
-    info = ServiceInfo(ZEROCONF_TYPE, zeroconf_name, host_ip_pton,
-                       hass.http.server_port, 0, 0, params)
-
-    zeroconf.register_service(info)
+    await zeroconf.register_service(info)
 
+    @callback
     def stop_zeroconf(event):
         """Stop Zeroconf."""
-        zeroconf.unregister_service(info)
-        zeroconf.close()
+        async def stop():
+            """Stop Zeroconf."""
+            await zeroconf.unregister_service(info)
+            await zeroconf.close()
+        hass.async_create_task(stop())
```

Comment:
We shouldn't create task, we should await it . Make `stop_zeroconf` async.

Reply:
Couldn't find a way for hass.bus.async_listen to run coroutines

Reply:
It does. All listeners are passed to `async_add_job` https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/core.py#L546-L547

Reply:
Great! I must have missed making it async def or something then... Will try tonight